# Ejercicios

Expresiones en Python, el tipo string.

## Ejercicio 1

Escribe el código Python que representa la siguiente expresión:

$$\frac{3}{4} + 2 * ( 5^4 - \frac{1}{2} )$$



In [1]:
x = 3/4
y = 5**4 - 1/2
x+2*y

1249.75

In [2]:
3/4+2*(5**4-1/2)

1249.75

Escribe una función en Python llamada __valor_expresion__ que calcule el valor de la expresión:

$$\frac{3}{4} + 2 * ( x^4 - \frac{1}{2} )$$

para un cierto valor de $x$. La función tiene un único argumento posicional de tipo numérico y devuelve el resultado de aplicar la expresión anterior.

In [6]:
def valor_expresion(x):
    resultado = 3/4+2*(x**4-1/2)
    return resultado

Ejecuta la función __valor_expresion__ para los valores 5 y 3. 

__Nota:__ 

* `valor_expresion(5)` devuelve como resultado `1249.75`
* `valor_expresion(3)` devuelve como resultado `161.75`

In [8]:
valor_expresion(5)

1249.75

In [9]:
valor_expresion(3)

161.75

## Ejercicio 2

Escribe una función en Python llamada __calcula_horas__ que calcule el número de horas completas hay en `m` minutos para un cierto valor de `m`. La función tiene un único argumento posicional de tipo entero (`int`) y devuelve el número de horas que también será entero.
Para resolverlo usa la división entera (`//`).

Ejecuta la función __calcula_horas__ para los valores 135 y 258. 

__Nota:__ 

* `calcula_horas(135)` devuelve como resultado `2`
* `calcula_horas(258)` devuelve como resultado `4`

In [10]:
def calcula_horas(x):
    horas = x//60
    return horas

In [11]:
calcula_horas(135)

2

In [12]:
calcula_horas(258)

4

## Ejercicio 3

Escribe una función en Python llamada __calcula_horas_minutos__ que calcule el número de horas completas y minutos que hay en `m` minutos para un cierto valor de `m`. La función tiene un único argumento posicional de tipo entero (`int`) y devuelve dos valores: el número de horas que también será entero y el número de minutos.
Para resolverlo usa la división entera (`//`) y la operación modulo (`%` resto de la división entera).

__Nota:__ 

* `calcula_horas_minutos(135)` devuelve como resultado `(2, 15)`
* `calcula_horas_minutos(258)` devuelve como resultado `(4, 18)`

In [13]:
def calcula_horas_minutos(a):
    h = a//60
    m = a%60
    return h,m

In [14]:
calcula_horas_minutos(135)

(2, 15)

In [15]:
calcula_horas_minutos(258)

(4, 18)

## Ejercicio 4

La función `upper` del tipo string (`str`) genera una nueva cadena de caracteres, donde todos los caracteres son mayúsculas.
Ejecuta la siguente celda para ver el efecto de la función `upper` al aplicarlo sobre la cadena `s`:

In [17]:
s = 'Hola, me llamo Iñigo Montoya, tú mataste a mi padre. ... Prepárate a morir'
s.upper()

'HOLA, ME LLAMO IÑIGO MONTOYA, TÚ MATASTE A MI PADRE. ... PREPÁRATE A MORIR'

Fíjate que `upper` no modifica la cadena (las cadenas son inmutables), sino que devuelve otra cadena diferente.

In [11]:
s    # la cadena s no cambia de valor

'Hola, me llamo Iñigo Montoya, tú mataste a mi padre. ... Prepárate a morir'

Para guardar el resultado de `upper` usamos la asignación:

In [12]:
t = s.upper()

Podemos  preguntar por el contenido de `t`:

In [13]:
t

'HOLA, ME LLAMO IÑIGO MONTOYA, TÚ MATASTE A MI PADRE. ... PREPÁRATE A MORIR'

* Genera una cadena `t`  igual que  `s`  pero en minúsculas (función `lower`). Calcula la longitud de `t`.
* Ejecuta la función  `split(',')`. Fíjate en lo que devuelve esa función. 
* Ejecuta la función  `title()`.
* Ejecuta la función  `replace(' ', '_')`. Fíjate en lo que devuelve esa función.
* Ejecuta `dir(str)` para listar las funciones disponibles para trabajar con cadenas.

In [18]:
t = s.lower()

In [19]:
t

'hola, me llamo iñigo montoya, tú mataste a mi padre. ... prepárate a morir'

In [20]:
len(t)

74

In [22]:
s.split(',')

['Hola',
 ' me llamo Iñigo Montoya',
 ' tú mataste a mi padre. ... Prepárate a morir']

In [23]:
s.title()

'Hola, Me Llamo Iñigo Montoya, Tú Mataste A Mi Padre. ... Prepárate A Morir'

In [24]:
s.replace(' ','_')

'Hola,_me_llamo_Iñigo_Montoya,_tú_mataste_a_mi_padre._..._Prepárate_a_morir'

In [26]:
dir(str)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

## Ejercicio 5

Es frecuente tener que convertir de unas unidades de medida a otras. Escribe expresiones para:
* pasar de kilogramos a gramos, dada una cantidad m que representa un peso en kilogramos,  dar su equivalente en gramos;
* pasar de kilogramos a libras (1kg = 2,20462 libras);
* pasar de grados Celsius a grados Fahrenheit (para  calcular cuántos grados Fahrenheit son X grados Celsius, tenemos que multiplicar los X grados Celsius por 1,8 y sumarle 32)



In [37]:
def kg_gramos(k):
    g=k*1000
    return g

In [28]:
def kg_libras(k):
    l=k*2.20462
    return l

In [29]:
def cel_fahr(c):
    f=c*1.8+32
    return f

In [40]:
kg_gramos (5)

5000

In [41]:
kg_libras (34)

74.95707999999999

In [42]:
cel_fahr (23)

73.4